crypto/test/Test/Pos/Crypto/Bi.hs

```diff
@@ -0,0 +1,294 @@
+                             deriveHDPassphrase, deterministicVssKeyGen, hash, packHDAddressAttr,
+                             redeemDeterministicKeyGen, toVssPublicKey)
+
+import           Test.Pos.Crypto.Gen
+import           Test.Pos.Crypto.TempHelpers (discoverGolden, discoverRoundTrip, eachOf,
+                                              goldenTestBi, roundTripsAesonBuildable,
+                                              roundTripsAesonShow, roundTripsBiBuildable,
+                                              roundTripsBiShow)
+
+--------------------------------------------------------------------------------
+-- ProtocolMagic
+--------------------------------------------------------------------------------
+
+roundTripProtocolMagicAeson :: Property
+roundTripProtocolMagicAeson = eachOf 1000 genProtocolMagic roundTripsAesonShow
+
+--------------------------------------------------------------------------------
+-- PublicKey
+--------------------------------------------------------------------------------
+
+golden_PublicKey :: Property
+golden_PublicKey = goldenTestBi pkey "test/golden/PublicKey"
+  where
+    Right pkey = PublicKey <$> xpub (getBytes 0 64)
+
+roundTripPublicKeyBi :: Property
+roundTripPublicKeyBi = eachOf 1000 genPublicKey roundTripsBiBuildable
+
+roundTripPublicKeyAeson :: Property
+roundTripPublicKeyAeson = eachOf 1000 genPublicKey roundTripsAesonBuildable
+
+--------------------------------------------------------------------------------
+-- SecretKey
+--------------------------------------------------------------------------------
+
+golden_SecretKey :: Property
+golden_SecretKey = goldenTestBi skey "test/golden/SecretKey"
+  where
+    Right skey = SecretKey <$> xprv (getBytes 10 128)
+
+roundTripSecretKeyBi :: Property
+roundTripSecretKeyBi = eachOf 1000 genSecretKey roundTripsBiBuildable
+
+--------------------------------------------------------------------------------
+-- Signature
+--------------------------------------------------------------------------------
+
+genUnitSignature :: Gen (Signature ())
+genUnitSignature = genSignature $ pure ()
+
+roundTripSignatureBi :: Property
+roundTripSignatureBi = eachOf 1000 genUnitSignature roundTripsBiBuildable
+
+roundTripSignatureAeson :: Property
+roundTripSignatureAeson = eachOf 1000 genUnitSignature roundTripsAesonBuildable
+
+--------------------------------------------------------------------------------
+-- Signed
+--------------------------------------------------------------------------------
+
+roundTripSignedBi :: Property
+roundTripSignedBi = eachOf 1000 genUnitSigned roundTripsBiShow
+  where
+    genUnitSigned = genSigned $ pure ()
+
+--------------------------------------------------------------------------------
+-- EncryptedSecretKey
+--------------------------------------------------------------------------------
+
+{-- This is disabled because we don't have an Eq instance for XPrv
```

Comment:
We might consider adding a comment here, something like:

```
Per discussion with Vincent, XPrv doesn’t have an Eq instance because we
can’t easily (and haven’t) verified that such an instance would run in constant
time and therefore not be vulnerable to timing attacks. 
```